File: Object_tracking_using_OpenCV.py
import cv2
import sys
 
(major_ver, minor_ver, subminor_ver) = (cv2.__version__).split('.')
 
if __name__ == '__main__' :
 
    # Установка трекера
    tracker_types = ['MEDIANFLOW' ,'KCF', 'MOSSE', 'CSRT']
    scan_num = len(tracker_types)

    for number in range(2,6):
        # Создание нового текстового файла
        log = open(f'Performance_test_crop_{number}.txt', 'w')
        log.write('\t\tFPS\tLOSS RATE\tLOSS TIME\tTrue/False\n')

        for j in range(scan_num):
            tracker_type = tracker_types[j]
            # Установка трекера
            if int(minor_ver) < 3:
                tracker = cv2.Tracker_create(tracker_type)
            else:
                if  tracker_type == 'KCF':
                    tracker = cv2.TrackerKCF_create()
                elif tracker_type == 'MEDIANFLOW':
                    tracker = cv2.legacy.TrackerMedianFlow_create()
                elif tracker_type == 'MOSSE':
                    tracker = cv2.legacy.TrackerMOSSE_create()
                elif tracker_type == "CSRT":
                    tracker = cv2.TrackerCSRT_create()
                else:
                    print("ERROR")
            # Установка доверительного трекера
            tracker_trust = cv2.TrackerCSRT_create()
            # Читать видео
            video = cv2.VideoCapture(f"crop_{number}.mp4")
        
            # Выход, если видео не открыто
            if not video.isOpened():
                print("Не удалось открыть видео")
                sys.exit()
            
            # Считывание первого кадра.
            ok, frame = video.read()
            if not ok:
                print('Невозможно прочитать видеофайл')
                sys.exit()
            
            # Определение начального ограничительного поля
            bbox = (287, 23, 86, 320)
        
            # выбор интересующей области на изображении
            if(j == 0):
                bbox0 = cv2.selectROI(frame, False)
            bbox = bbox0
            # Инициализация трекера с первым кадром и ограничительной рамкой
            ok = tracker.init(frame, bbox)
            ok_trust = tracker_trust.init(frame, bbox)

            i=0                         # Число итераций
            fps0=0                      # Усредненное 
            f = 0                       # Флаг потери обЪекта
            #f_trust = 0                 # 
            time_loss = (str)(None)     # Впремя потери обЪекта (вычисляется не корректно)
            LOSS_RATE = 0               # Число потерь обЪекта
            false_number = 0            # Число отличий от доверительного трекера
            true_number = 0             # Число совпадений с доверительным трекером

            #Настройка сохранения видео файла
            fourcc = cv2.VideoWriter_fourcc('m', 'p', '4', 'v')
            out = cv2.VideoWriter(f'output_{number}_{tracker_type}.mp4', fourcc, 30.0, (1920, 1080))

            timer0 = cv2.getTickCount()
        
            while True:
                i+=1
                # Считывание нового кадра
                ok, frame = video.read()
                if not ok:
                    fps0 = fps0/(i-1)
                    print("среднее число FPS "+(str)(fps0))
                    break
                
                # Запуск таймера
                timer = cv2.getTickCount()
        
                # Обновление трекера
                ok, bbox = tracker.update(frame)
        
                # Рассчитать количество кадров в секунду (FPS)
                fps = cv2.getTickFrequency() / (cv2.getTickCount() - timer)
                fps0 += int(fps)

                # Обновление доверительного трекера
                ok_trust, bbox_trust = tracker_trust.update(frame)

                # Отслеживание точности трекера
                if(ok and ok_trust):
                    if(bbox[0] <= (bbox_trust[0] + bbox_trust[2]/2) and (bbox[0] + bbox[2]) >= (bbox_trust[0] + bbox_trust[2]/2)):
                        if(bbox[1] <= (bbox_trust[1] + bbox_trust[3]/2) and (bbox[1] + bbox[3]) >= (bbox_trust[1] + bbox_trust[3]/2)):
                            true_number += 1
                            #f_trust = 0
                        else:
                            false_number += 1
                            # if(f_trust == 0):
                            #     LOSS_RATE += 1
                            #     f_trust = 1
                    else:
                        false_number += 1
                        # if(f_trust == 0):
                        #     LOSS_RATE += 1
                        #     f_trust = 1
        
                # Нарисовать ограничительную рамку
                if ok:
                    # Отслеживание успеха
                    if(f == 1):
                        f = 0
                    p1 = (int(bbox[0]), int(bbox[1]))
                    p2 = (int(bbox[0] + bbox[2]), int(bbox[1] + bbox[3]))
                    cv2.rectangle(frame, p1, p2, (255,0,0), 2, 1)
                    cv2.putText(frame, "x " + str(int(bbox[0] + bbox[2]/2)) + "; y " + str(int(bbox[1] + bbox[3]/2)), (100,80), cv2.FONT_HERSHEY_SIMPLEX, 0.75,(0,0,255),2)
                else :
                    # Сбой отслеживания
                    if(f == 0):
                        LOSS_RATE += 1
                        time_loss = (str)((cv2.getTickCount() - timer0)/cv2.getTickFrequency())
                        print("среднее число FPS "+(str)(fps0/i))
                        f=1
                    cv2.putText(frame, "Обнаружен сбой отслеживания", (100,80), cv2.FONT_HERSHEY_SIMPLEX, 0.75,(0,0,255),2)
        
                # Отображение типа трекера на рамке
                cv2.putText(frame, tracker_type + " Tracker", (100,20), cv2.FONT_HERSHEY_SIMPLEX, 0.75, (50,170,50),2)
            
                # Отображение FPS на кадре
                cv2.putText(frame, "FPS : " + str(int(fps)), (100,50), cv2.FONT_HERSHEY_SIMPLEX, 0.75, (50,170,50), 2)
        
                # Результат отображения
                cv2.imshow("Tracking", frame)
                out.write(frame)
        
                # Выход при нажатии ESC
                k = cv2.waitKey(1) & 0xff
                if k == 27 : break

            video.release()
            out.release()
            cv2.destroyAllWindows()
            
            # Добавление данных отслеживания в новую строку таблицы
            log.write(str(tracker_type)+'\t')
            log.write(str((int)(fps0))+'\t')          
            log.write(str(LOSS_RATE)+'\t')     
            log.write(str(time_loss)+'\t')     
            log.write((str)(round(true_number/(i-1)*100, 2))+"/"+(str)(round(false_number/(i-1)*100, 2))+'\n')    

        # Сохранение данных в текстовый файл
        log.close()

File: Object_tracking_using_OpenCV.cpp
#include <opencv2/opencv.hpp>
#include <opencv2/tracking.hpp>
#include <opencv2/tracking/tracking_legacy.hpp>
#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace cv;
using namespace std;

int main(int argc, char **argv)
{
    // выбор интересующей области на изображении 
    Rect2d bbox0(287, 23, 86, 320);
    Rect2d bbox(287, 23, 86, 320);
    Rect2d bbox_trust(287, 23, 86, 320); 
    // List of tracker types in OpenCV 3.4.1
    string trackerTypes[4] = {"MEDIANFLOW", "KCF", "MOSSE", "CSRT"};
    int N_len = 4;
    // vector <string> trackerTypes(types, std::end(types));
    for(int number=2; number < 6; number++)
    {
        // Создание нового текстового файла
        //описывает поток для записи данных в файл
        ofstream out_text;
        //открываем файл в режиме записи,
        //режим ios::out устанавливается по умолчанию
        out_text.open("Performance_test_crop_"+to_string(number)+".txt", ios::out);
        out_text << "\t\tFPS\tLOSS RATE\tLOSS TIME\tTrue/False" << endl;
        for(int j=0; j<N_len; j++)
        {
            // Create a tracker
            string trackerType = trackerTypes[j];
        
            Ptr<cv::legacy::Tracker> tracker;
            Ptr<cv::legacy::Tracker> tracker_trust;
        
            #if (CV_MINOR_VERSION < 3)
            {
                tracker = Tracker::create(trackerType);
            }
            #else
            {
                if (trackerType == "KCF")
                    tracker = cv::legacy::TrackerKCF::create();
                if (trackerType == "MEDIANFLOW")
                    tracker = cv::legacy::TrackerMedianFlow::create();
                if (trackerType == "MOSSE")
                    tracker = cv::legacy::TrackerMOSSE::create();
                if (trackerType == "CSRT")
                    tracker = cv::legacy::TrackerCSRT::create();
            }
            #endif
            // Установка доверительного трекера
            tracker_trust = legacy::TrackerCSRT::create();
            // Read video
            VideoCapture video("crop_"+to_string(number)+".mp4");
            
            // Exit if video is not opened
            if(!video.isOpened())
            {
                cout << "Could not read video file" << endl; 
                return 1; 
            } 
        
            // Read first frame 
            Mat frame; 
            bool ok = video.read(frame); 
    
            if(j == 0)
                bbox0 = selectROI(frame, false); 
        
            // Uncomment the line below to select a different bounding box 
            bbox = bbox0; 
            // Display bounding box. 
            //rectangle(frame, bbox, Scalar( 255, 0, 0 ), 2, 1 ); 
        
            // Uncomment the line below to select a different bounding box 
            bbox_trust = bbox0; 

            int i=0;                         // Число итераций
            int fps0=0;                      // Усредненное 
            int f = 0;                       // Флаг потери обЪекта
            string time_loss = "None";  // Впремя потери обЪекта (вычисляется не корректно)
            int LOSS_RATE = 0;               // Число потерь обЪекта
            int false_number = 0;            // Число отличий от доверительного трекера
            int true_number = 0;             // Число совпадений с доверительным трекером
        
            imshow("Tracking", frame); 
            tracker->init(frame, bbox);
            tracker_trust->init(frame, bbox_trust);

            int fourcc = VideoWriter::fourcc('m', 'p', '4', 'v');
            VideoWriter outputVideo("output_cpp_"+to_string(number)+"_"+trackerType+"_type}.mp4", fourcc, 30.0, Size(1920, 1080)); // For writing the video

            double timer0 = (double)getTickCount();
            
            while(video.read(frame))
            {     
                i++;
                // Start timer
                double timer = (double)getTickCount();
                
                // Update the tracking result
                bool ok = tracker->update(frame, bbox);
                
                // Calculate Frames per second (FPS)
                float fps = getTickFrequency() / ((double)getTickCount() - timer);
                fps0 += int(fps);

                // Обновление доверительного трекера
                bool ok_trust = tracker_trust->update(frame, bbox_trust);

                // Отслеживание точности трекера
                if(ok and ok_trust)
                {
                    if(bbox.x <= (bbox_trust.x + bbox_trust.width/2) and (bbox.x + bbox.width) >= (bbox_trust.x + bbox_trust.width/2))
                    {
                        if(bbox.y <= (bbox_trust.y + bbox_trust.height/2) and (bbox.y + bbox.height) >= (bbox_trust.y + bbox_trust.height/2))
                            true_number ++;
                        else
                            false_number ++;
                    }        
                    else
                        false_number ++;
                }        
                
                if (ok)
                {
                    // Отслеживание успеха
                    if(f == 1)
                    {
                        f = 0;
                    }
                    // Tracking success : Draw the tracked object
                    rectangle(frame, bbox, Scalar( 255, 0, 0 ), 2, 1 );
                    putText(frame, "x: "+ to_string(int(bbox.x + bbox.width/2)) + "; y " + to_string(int(bbox.y + bbox.height/2)), Point(100,80), FONT_HERSHEY_SIMPLEX, 0.75, Scalar(0,0,255),2);
                }
                else
                {
                    if(f == 0)
                    {
                        LOSS_RATE ++;
                        time_loss = to_string((double)getTickCount() - timer0);
                        f = 1;
                    }
                    // Tracking failure detected.
                    putText(frame, "Tracking failure detected", Point(100,80), FONT_HERSHEY_SIMPLEX, 0.75, Scalar(0,0,255),2);
                }
                
                // Display tracker type on frame
                putText(frame, trackerType + " Tracker", Point(100,20), FONT_HERSHEY_SIMPLEX, 0.75, Scalar(50,170,50),2);
                
                // Display FPS on frame
                putText(frame, "FPS : " + to_string(int(fps)), Point(100,50), FONT_HERSHEY_SIMPLEX, 0.75, Scalar(50,170,50), 2);
        
                // Display frame.
                imshow("Tracking", frame);
                outputVideo.write(frame);
                
                // Exit if ESC pressed.
                int k = waitKey(1);
                if(k == 27)
                {
                    break;
                }
            }
            video.release();
            outputVideo.release();
            destroyAllWindows();
            fps0 = fps0/i;
            out_text << trackerType +"\t";
            out_text << to_string((int)(fps0))+"\t";
            out_text << to_string(LOSS_RATE)+"\t\t";
            out_text << time_loss +"\t";
            out_text << to_string(round(true_number/(i-1)*10000)/100)+"/"+to_string(round(false_number/(i-1)*10000)/100) << endl;
        }
        out_text.close(); 
    }
}